crates/client/commitment-state-diff/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+                        Err(e) => {
+                            log::error!(
+                                "Block with substrate hash `{block_hash}` skiped. Failed to compute commitment state \
+                                 diff: {e}",
+                            );
+
+                            return Poll::Pending;
+                        }
+                    }
+                }
+
+                // The stream has been close, we close too.
+                // This should not happen tho
+                Poll::Ready(None) => return Poll::Ready(None),
+            }
+        }
+
+        // At this point self_as_mut.commitment_state_diff.is_some() == true
+        // State 2
+        match self_as_mut.tx.poll_ready(cx) {
+            // Channel is ready, we send
+            Poll::Ready(Ok(())) => {
+                // Safe to unwrap cause we already handle the `None` branch
+                let msg = self_as_mut.msg.take().unwrap();
+                // Safe to unwrap because channel is ready
+                self_as_mut.tx.start_send(msg).unwrap();
+
+                Poll::Ready(Some(()))
+            }
+
+            // Channel is full, we wait
+            Poll::Pending => Poll::Pending,
+
+            // Channel receiver have been drop, we close.
+            // This should not happen tho
+            Poll::Ready(Err(e)) => {
+                log::error!("CommitmentStateDiff channel reciever have been droped: {e}");
+                Poll::Ready(None)
+            }
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+enum BuildCommitmentStateDiffError {
+    #[error("failed to interact with substrate header backend")]
+    SubstrateHeaderBackend(#[from] sp_blockchain::Error),
+    #[error("block not found")]
+    BlockNotFound,
+    #[error("digest log not found")]
+    DigestLogNotFound(#[from] mp_digest_log::FindLogError),
+}
+
+fn build_commitment_state_diff<B: BlockT, C, H>(
+    client: Arc<C>,
+    storage_notification: StorageNotification<B::Hash>,
+) -> Result<(BlockHash, CommitmentStateDiff), BuildCommitmentStateDiffError>
+where
+    C: ProvideRuntimeApi<B>,
+    C::Api: StarknetRuntimeApi<B>,
+    C: HeaderBackend<B>,
+    H: HasherT,
+{
+    let starknet_block_hash = {
+        let header = client.header(storage_notification.block)?.ok_or(BuildCommitmentStateDiffError::BlockNotFound)?;
+        let digest = header.digest();
+        let block = mp_digest_log::find_starknet_block(digest)?;
+        block.header().hash::<H>().into()
+    };
+
+    let mut commitment_state_diff = CommitmentStateDiff {
+        address_to_class_hash: Default::default(),
+        address_to_nonce: Default::default(),
+        storage_updates: Default::default(),
+        class_hash_to_compiled_class_hash: Default::default(),
+    };
+
+    for (_prefix, full_storage_key, change) in storage_notification.changes.iter() {
+        // The storages we are interested in here all have longe keys
```

Comment:
```suggestion
        // The storages we are interested in here all have long keys
```

Could you expand on this ?

Reply:
done